deep_qa/training/multi_gpu.py

```diff
@@ -1,88 +1,128 @@
-from keras.layers import concatenate
-from keras.layers.core import Lambda
-import keras.backend as K
+from typing import Callable
+import os
+from copy import deepcopy
+
 import tensorflow
+import keras.backend as K
 
+from .train_utils import pin_variable_device_scope, average_gradients
 from .models import DeepQaModel
+from .step import Step
+from ..common.params import Params, ConfigurationError
 
 
-def pin_variable_device_scope(device, variable_device="/cpu:0"):
+def compile_parallel_model(model_builder: Callable[[any], DeepQaModel],
+                           compile_arguments: Params) -> DeepQaModel:
     """
-    Tensorflow device scopes can take functions which give a device
-    for a given op in the graph. Here, we use the device that is
-    passed to the scope *unless* the operation which is being created
-    in the graph is a Variable creation op; in this case, we place it
-    on the cpu.
-    """
-    def _assign(graph_op):
-        node_def = graph_op if isinstance(graph_op, tensorflow.NodeDef) else graph_op.node_def
-        if node_def.op in ['Variable', 'VariableV2']:
-            return variable_device
-        else:
-            return device
-    return _assign
+    This function compiles a multi-gpu version of your model. This is done using data
+    parallelism, by making N copies of the model on the different GPU, all of which
+    share parameters. Gradients are updated synchronously, using the average gradient
+    from all of the outputs of the various models. This effectively allows you to scale
+    a model up to batch_sizes which cannot fit on a single GPU.
 
+    This method returns a "primary" copy of the model, which has had it's training
+    function which is run by Keras overridden to be a training function which trains
+    all of the towers of the model. The other towers never have their training functions
+    initialised or used and are completely hidden from the user. The returned model
+    can be serialised in the same way as any other model and has no dependency on
+    multiple gpus being available when it is loaded.
+
+    Note that in the multiple gpu case, the model_builder function will be called multiple times
```

Comment:
Calling this function _is_ the "multiple GPU case", so you can probably simplify this.